### PR TITLE
perf: make more items stackable

### LINKED ISSUE #3487: ☂️ make rags lag less
## Summary

Make items hoarded a lot stackable to reduce lag

## Process

1. change `type` to `AMMO`
2. give it an `ammo_type` of `components`
3. give it an `count` of `1`
4. additionally add `abstract` and use `copy-from`

```[tasklist]
### Item Ids
- [x] rag
- [x] rag_bloody
- [x] disinrag
- [x] nylon
- [x] superglue
- [x] plastic_chunk
- [x] string_6
- [x] 2x4
- [x] nomex
- [x] felt_patch
- [x] leather
- [x] fur
- [x] faux_fur
- [x] neoprene
```

```[tasklist]
### Now that they're stacked
- [ ] Simplify ropes?
- [x] Write a script so i don't have to manually do this stuff
```

### PR DESCRIPTION
**Pull request by scarf005**
## Purpose of change

- resolve #3487

## Describe the solution

1. wrote a script to migrate items to ~~AMMO~~ stackable
2. converted items i could think of
3. ~~rags are heavily buffed, weighs only 10g/20ml~~ (should be a separate balance PR)
4. `TOOL`s are now `GENERIC` 
5. parts that were also used as vehicle parts had to be removed as they where the culprit for breaking vehicle tests. i suspect vehicle drag and efficiency calculate does not take account of count-by-charges

## Describe alternatives you've considered

make everything stackable?

## Testing

haven't tested disassembly and crafting, so weird stuff might happen.

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/8fd48737-2589-44d2-9e68-b1c922dfb08e

interestingly, `use_action` works as before.

## Additional context

context: https://discord.com/channels/830879262763909202/830916329053487124/1166006575664738307

### CODE REVIEW

**Comment by olanti-p** [comment]
`AMMO` is for gun and tool ammo. If you want generic items to be stackable, use `stackable`.

**Comment by scarf005** [comment]
ah, will fix the script (tomorrow)

**Comment by scarf005** [comment]
```
 DEBUG    : warnings for type disinrag:
invalid stack_size 0 on type using charges


 FUNCTION : void Item_factory::check_definitions() const
 FILE     : /home/scarf/repo/cata/Cataclysm/src/item_factory.cpp
 LINE     : 1526
 VERSION  : BN 9fd976a84ee6
```
oww

**Comment by olanti-p** [comment]
Test failures seem to be related to the change, no other PR has them.

**Comment by scarf005** [comment]
i've reduced rag weight and volume, so vehicle distance travelled test was failing.

**Comment by chaosvolt** [comment]
Hmm, this is gonna run into the same issue as https://github.com/cataclysmbnteam/Cataclysm-BN/pull/3414 in opening up a lot of potential material duplication exploits. :/

**Comment by scarf005** [comment]
probably this indicates we need a saner recipe and uncrafts system, but anyways will rollback rag weights

**Comment by chaosvolt** [comment]
Either way is fine I guess, the fix PR is kinda slow going since it's absolute hell to work with each item so do whichever works. XD

**Comment by scarf005** [comment]
> the fix PR is kinda slow going since it's absolute hell to work with each item

well that sounds painful, let's don't, you deserve better

**Comment by chaosvolt** [comment]
Eh, items and recipes kinda need a big cleanup either way and have needed it ever since salvage-by-weight was merged, just still unsure which direction it should go in.

File: data/json/items/generic.json [hunk 1]
```diff
@@ -1677,16 +1685,37 @@
     "to_hit": -6
   },
   {
+    "abstract": "rag_abstract",
     "type": "GENERIC",
-    "id": "rag_bloody",
+    "name": { "str": "rag" },
+    "description": "Rag",
```
Comment (Coolthulhu) [inline data/json/items/generic.json:1691]: A bit short. What's the most common type of description for an abstract?

**Comment by scarf005** [comment]
currently vehicle tests fail because apparently item weights are calculated differently for stacks.

**Comment by chaosvolt** [comment]
Side note: just in case, test if having rags be stackable affects them being usable for raw materials, both for sewing repairs and for modifying with clothing_mods (now that we have a vanilla clothing mod that uses rags). I recall in the past running in an issue where I tried to use a stacking item as the material for a tailor's kit mod (was either bones or rocks) and it always said it didn't have any available.

File: data/json/items/generic.json (hunk 1 above)
Comment (scarf005) [reply data/json/items/generic.json]: renamed to `rag_abstract`.

**Comment by scarf005** [comment]
found the curprit, it was `glass_sheet`.

**Comment by scarf005** [comment]
a hecc, count by charges don't work with sewing kit
i'll make them check by their charges

**Comment by chaosvolt** [comment]
RIP. Did it only affect using them as reagents to add `clothing_mod`s or did it also affect using them for item repair?

**Comment by scarf005** [comment]
https://github.com/cataclysmbnteam/Cataclysm-BN/blob/7143dfd4038c3b15c356b080ed21e394dc402124/src/visitable.cpp#L1109-L1115

haven't tested but probably both as `visitable<T>::amount_of` doesn't seem to take charges into account

**Review by chaosvolt — CHANGES_REQUESTED** [review]
1. Compiled and load-tested.
2. Took down some curtains and disassembled the sheet, got 20 rags as expected.
3. Cut up the other rags, message is a lil weird (see below) but yield was as expected, and amount spawned matches what it said spawned.
4. Uncrafted the long string into short strings then thread, correctly got 300 thread out of it.
5. Repeated on another curtain and hauled the two piles of items together, items stack up correctly.
6. Made a french maid dress, it correctly used the 30 rags it said it should.
7. Reinforced some items, it consumes rags as expected.
6. Checked via yeeting remaining rags away that a sheet says it takes 2 rags per repair operation, destroyed another sheet and confirmed it takes the stated number of rags when sewing.
7. Enlarge-ified my shirt, it correctly consumed 2 rags.
8. Constructed and deconstructed a wire fence, it correctly takes 2, and yields 2 wires.
9. Bolt-cut the wire fence too, yields 2 wires as expected.
image

10. And the bad bit: Bloodied myself up and used a rag, it generates a blood-stained rag but doesn't consume the rag.
image

**Review by chaosvolt — CHANGES_REQUESTED** [review]
Bonus WTF: splintered woods seem to not work right for fuel in a firewood source.
image
Except sometimes they do? what in the goddamn
image

**Comment by scarf005** [comment]
image

> Bonus WTF: splintered woods seem to not work right for fuel in a firewood source.

strange, i cannot reproduce this one.

**Comment by chaosvolt** [comment]
Tested it again and I still don't understand the firewood source behavior. What I did, in order:
1. Created new world.
2. Took down a curtain.
3. Used the stick to smash a bunch of benches and some lockers.
4. Made a brazier, piled all the wood in a spot next to it.
5. Defined a firewood source from the construction menu, on the pile I made.
6. Examined the brazier and lit a fire, it immediately worked first time.
7. Moved the planks off the pile, removed the firewood source, and re-added it again to the pile that now has just splintered woods on a freshly-constructed firewood source.

I have a save from right this moment of it having not worked:
[Offerman.zip](https://github.com/cataclysmbnteam/Cataclysm-BN/files/13764969/Offerman.zip)

After I saved I try the brazier, it works. I quit without saving, load save, try again and this time it doesn't work. So whatever the hell this is, it's literally random if it works or not. I figured maybe it was a wind effect because my earlier test was outside and it was just being a shit and not printing a message about wind causing fires to not work or something, but this save's test is purely indoors.

**Comment by scarf005** [comment]
https://github.com/cataclysmbnteam/Cataclysm-BN/blob/7143dfd4038c3b15c356b080ed21e394dc402124/src/item.cpp#L8300-L8305

I'm suspecting that this line is making random failure.

**Comment by scarf005** [comment]
now it should work.

**Review by chaosvolt — APPROVED** [review]
Yep, that seems to have done it. :3

**Pull request merged**